doc/rfcs/003-quickly-collaborate-with-coworkers-and-friends.md

```diff
@@ -0,0 +1,146 @@
+
+##### What other approaches have been considered and what is the rationale for not choosing them?
+
+- Teletype could allow you to enter a person's GitHub username (or their email address) to invite them to your portal. This would remove the need for sharing a URL via a third-party service in order to collaborate, and it would reduce the need to provide a list of past collaborators (i.e., you could just enter a username each time you want to collaborate). However, it would make it possible for any GitHub user to cause invitations to appear inside your Atom instance. This introduces a vector for harassment, so we're avoiding this approach.
+- Teletype could allow you to invite other users that you're already associated with in some way (e.g., fellow collaborators on a GitHub repository, fellow members of a GitHub organization or team). This would remove the need for sharing a URL via a third-party service in order to collaborate, and it would reduce the need to provide a list of past collaborators, and it would reduce the harassment vector described in the previous bullet. However, it would introduce tradeoffs that we'd prefer to avoid:
+    - Teletype would need additional permissions (i.e., [OAuth scopes](https://developer.github.com/apps/building-oauth-apps/scopes-for-oauth-apps/#available-scopes)) to fetch the list of users you're associated with. (Today, Teletype  uses a scopeless token and has no access to your private information.)
+    - Some users belong to organizations with thousands of members. Just because you're a member of the same organization as someone else doesn't mean that you're comfortable seeing portal invitations from them.
+
+##### What is the impact of not doing this?
+
+People will collaborate less often. Given the additional steps needed to start collaborating, there will be more instances where people decide that it's not worth the effort (i.e., people will enjoy rich collaboration less often).
+
+## Unresolved questions
+
+##### What unresolved questions do you expect to resolve through the RFC process before this gets merged?
+
+- If I invite a past collaborator to join my portal, and that person has multiple Atom windows open, can we show the invitation only in the frontmost window?
+- If I invite a past collaborator to join my portal, and that person has Atom windows open on multiple computers, are we OK with showing the invitation on each computer?
+- Where will the list of past collaborators appear in the UI? How do I show/hide this list?
```

Comment:
Could the list still be part of the "popup" (click on the Teletype icon in the status bar)? The list of past collaborators and the list of current collaborators could be the same thing. Just that current collaborators would move to the top and instead of a "Invite" button, they would have a "Follow cursor" button.

An alternative could be a separate panel/dock item, but that feels a bit overkill with the current features. Having the list in the popup is nice because it auto-closes when done inviting and doesn't compete with the tree-view or so. I'll make a mockup so we can see how that feels.

Reply:
Mockup of a combined current and past collaborator list:

![8](https://user-images.githubusercontent.com/378023/37592124-1a551604-2bb1-11e8-8ea4-e75c63a4e04e.png)

- After sharing your portal, you see the list of past contributors.
- Click on "Invite" to invite someone.
- The "Invite" button turns then into a "Invitation sent" status and that collaborator moves to the top.
- After a collaborator joins, the "Invitation sent" status turns into a "Follow" button. It's the same as clicking an avatar in the editor to follow someone's cursor.
- If the invitee doesn't accept the invitation, the host see a "Invitation rejected" status.
- Alternatively or in addition to clicking the "Invite" buttons, a host can also copy the URL at the bottom and share anywhere.
- When the host clicks on the "Share" toggle, the portal is closed again.

And a click through:

![portal](https://user-images.githubusercontent.com/378023/37592135-26898bd0-2bb1-11e8-9d6c-604b1159c157.gif)

There are still a few open questions, like should the list already be visible, even without sharing? Or can you re-invite someone when they don't join or reject. But we can get to those if we like this direction.

Reply:
@simurai: Thanks for this mockup. I really dig where this is going. :zap::bow:

> There are still a few open questions...But we can get to those if we like this direction.

I like this direction. 😄 Would you mind taking a stab at those open questions?

A few questions for you about the current mockup:

1. I think we'll still want to provide a way to join a portal via the UI (https://github.com/atom/teletype/pull/344#discussion_r175438235). What would you think about keeping the existing "Join a Portal" option at the bottom of the popup?

    <img width="314" src="https://user-images.githubusercontent.com/2988/37612836-b3f2d556-2b7c-11e8-8857-b08266c852e3.png">

2. Can you share a mockup for removing someone from your list of past collabortors?

3. It's rare (but possible) to host a portal while also being a guest of a portal:

    <img width="311" src="https://user-images.githubusercontent.com/2988/37613326-1b819a9e-2b7e-11e8-8ee2-7f17ee5b0916.png">

    Can you share a mockup that shows this scenario?

4. Assuming that we want to get the initial release in people's hands [without making people wait for us to implement sorting, filtering, favoriting](https://github.com/atom/teletype/blob/e35483b1fa69c3ef1ef56f0c437000d4dcda934b/doc/rfcs/003-quickly-collaborate-with-coworkers-and-friends.md#out-of-scope), can you offer any thoughts on how the UI should gracefully handle long lists of collaborators in the meantime?

Reply:
1. Ok, yeah, let's keep the "Join a portal" button
2. There could be an X on right of each past collaborator, but that feels a bit too aggressive and can easily be clicked by accident. So how about have a right-click context-menu? Then we can also explain the action in words. -> "Remove collaborator". Sorry, @as-cii :stuck_out_tongue_winking_eye: 
  ![image](https://user-images.githubusercontent.com/378023/37639861-67778f8e-2c56-11e8-994f-428802fa9191.png)
3. Oh.. right. And it seems as a guest, you can even join multiple times. You have to be a very good multi-tasker, though. Ok, to not take up much more space, we could keep joined guests as is -> Only show the avatar next to the host. I think that's fine, because you probably know who the guests are since you just invited them. So it could look like this:
  ![host guest 2](https://user-images.githubusercontent.com/378023/37640529-6a5ddaac-2c59-11e8-82f5-9ee4c2ed42a5.png)
4. The list of past collaborators has a `max-height`, if the list is longer, you can scroll. We can play with the number, but maybe so that 5 collaborators fit? The scrolling area includes sent and rejected invitations. Then maybe stop rendering at like 100 collaborators, just as a safeguard. If your list is larger than that, you would need some sort of search/filter anyways.
  ![host](https://user-images.githubusercontent.com/378023/37642043-1d66797e-2c5f-11e8-8def-01f063888979.png)

Reply:
Now about those other open questions:

- **Should the list already be visible, even without sharing?** I guess? If you click on "Invite" before starting to share, it would start share automatically.
  ![image](https://user-images.githubusercontent.com/378023/37643605-6947af84-2c64-11e8-94b6-849e81fd8cd0.png)
- **Can you invite someone if they are offline?** Yes, absolutely. Having to check back every few minutes to see if they are online would be a pain. As soon as the invitee opens Atom, the notification appears. Seeing that someone is offline is more FYI that it might take a while till that collaborator joins.
- **Can you re-invite someone when they don't join or reject?** No. There must be reasons why someone doesn't join, no need to bug them even more. If you REALLY have to, you could stop and start sharing, and then you can invite again. Or probably better, send them a chat message or so.

Reply:
Thank you, @simurai! 💟

I've incorporated these mockups and scenarios into the RFC.